concordium-contracts-common/src/schema_json.rs

```diff
@@ -40,8 +42,133 @@ pub enum JsonError {
     ParseDurationError(#[from] ParseDurationError),
     #[error("{0}")]
     ParseTimestampError(#[from] ParseTimestampError),
+    /// Trace leading to the original [JsonError].
+    #[error("{field} -> {error}")]
+    TraceError {
+        field: String,
+        json:  &'a serde_json::Value,
+        error: Box<JsonError<'a>>,
+    },
 }
 
+impl<'a> JsonError<'a> {
+    /// Wraps a [JsonError] in a [JsonError::TraceError], providing a trace to
+    /// the origin of the error.
+    fn add_trace(self, field: String, json: &'a serde_json::Value) -> Self {
+        JsonError::TraceError {
+            field,
+            json,
+            error: Box::new(self),
+        }
+    }
+
+    /// Gets the underlying error of a [JsonError::TraceError]. For any other
+    /// variant, this simply returns the error itself.
+    pub fn get_error(&self) -> &Self {
+        if let JsonError::TraceError {
+            error,
+            ..
+        } = self
+        {
+            return error.get_error();
+        }
+
+        self
+    }
+
+    /// Prints a formatted error message for variant. [JsonError::TraceError]
+    /// supports printing a verbose form including a more detailed
+    /// description of the error stack, which is returned if `verbose` is
+    /// set to true.
+    pub fn print(&self, verbose: bool) -> String {
+        match self {
+            JsonError::TraceError {
+                field,
+                json,
+                error,
+            } if verbose => {
+                let formatted_json =
+                    serde_json::to_string_pretty(json).unwrap_or(format!("{}", json));
+                format!("{}\nIn {} of {}", error.print(verbose), field, formatted_json)
+            }
+            _ => format!("{}", self),
+        }
+    }
+}
+
+/// Wrapper around a list of bytes to represent data which failed to be
+/// deserialized into schema type.
+#[derive(Debug, Clone)]
+pub struct ToJsonErrorData(Vec<u8>);
+
+impl From<Vec<u8>> for ToJsonErrorData {
+    fn from(value: Vec<u8>) -> Self { Self(value) }
+}
+
+impl Display for ToJsonErrorData {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        self.0.iter().map(|b| write!(f, "{:02x?}", b)).collect()
+    }
+}
+
+/// Represents errors occuring while deserializing to schema JSON format
+#[derive(thiserror::Error, Debug, Clone)]
+pub enum ToJsonError<'a> {
+    /// JSON formatter failed to represent value.
+    #[error("Failed to format as JSON")]
+    FormatError {},
+    /// Failed to deserialize data to type expected from schema.
+    #[error("Failed to deserialize {schema:?} from position {position} of bytes {data}")]
+    DeserialError {
+        position: u32,
+        schema:   &'a Type,
+        data:     ToJsonErrorData,
```

Comment:
It looks like in a number of cases you could provide some details here, for example "integer overflow".

Is there a reason for not adding an error message string here?